<a href="https://colab.research.google.com/github/Tambo-Academy/Deep-Learning/blob/main/Dealing_with_BERT_Overfitting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.


References:
1. [AUBER: Automated BERT regularization](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8238198/)
2. [Why transformer overfit quickly? how to solve it?](https://discuss.huggingface.co/t/why-transformer-overfit-quickly-how-to-solve-it/1842)
3. [Dropout Regularization to Handle Overfitting in Deep Learning Models](https://www.section.io/engineering-education/dropout-regularization-to-handle-overfitting-in-deep-learning-models/)
4. [Transfer Learning for NLP: Fine-Tuning BERT for Text Classification](https://www.analyticsvidhya.com/blog/2020/07/transfer-learning-for-nlp-fine-tuning-bert-for-text-classification/)
5. [How to Fine-Tune BERT for Text Classification?](http://cips-cl.org/static/anthology/CCL-2019/CCL-19-141.pdf)
6. [Handling overfitting to proper nouns in BERT-based transformer models](https://www.reddit.com/r/MachineLearning/comments/jcxhb5/d_handling_overfitting_to_proper_nouns_in/)
7. [Handling overfitting in deep learning models.ipynb](https://github.com/Tambo-Academy/TwitterUSAirlineSentiment/blob/master/source/Handling%20overfitting%20in%20deep%20learning%20models.ipynb)
8. [Overfitting in Huggingface's TFBertForSequenceClassification](https://datascience.stackexchange.com/questions/76527/overfitting-in-huggingfaces-tfbertforsequenceclassification)

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv("./sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
199  0.0116  0.0744  0.0367  0.0225  0.0076  0.0545  0.1110  0.1069  0.1708   
174  0.0191  0.0173  0.0291  0.0301  0.0463  0.0690  0.0576  0.1103  0.2423   
31   0.0084  0.0153  0.0291  0.0432  0.0951  0.0752  0.0414  0.0259  0.0692   
129  0.1371  0.1226  0.1385  0.1484  0.1776  0.1428  0.1773  0.2161  0.1630   
126  0.0715  0.0849  0.0587  0.0218  0.0862  0.1801  0.1916  0.1896  0.2960   

         9   ...      51      52      53      54      55      56      57  \
199  0.2271  ...  0.0141  0.0103  0.0100  0.0034  0.0026  0.0037  0.0044   
174  0.3134  ...  0.0040  0.0136  0.0137  0.0172  0.0132  0.0110  0.0122   
31   0.1753  ...  0.0236  0.0114  0.0136  0.0117  0.0060  0.0058  0.0031   
129  0.2067  ...  0.0118  0.0129  0.0344  0.0065  0.0067  0.0022  0.0079   
126  0.4186  ...  0.0153  0.0121  0.0096  0.0196  0.0042  0.0066  0.0099   

         58      59  60  
199  0.0057  0.0035   M  
174  0.0114  0.0068   M  
31   0.0072  0.0045   R  
129  0.0146  0.0051   M  
126  0.0083  0.0124   M  

[5 rows x 61 columns]

In [ ]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [ ]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
59   1
101  0
184  0
141  0
191  0

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

# Using Deep Learning Model

## Model without Dropout Layer

In [ ]:
import tensorflow as tf
from tensorflow import keras

In [ ]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6895 - accuracy: 0.5321
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6669 - accuracy: 0.5833
Epoch 3/100
20/20 [==============================] - 0s 1ms/step - loss: 0.6512 - accuracy: 0.6474
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6189 - accuracy: 0.6923
Epoch 5/100
20/20 [==============================] - 0s 1ms/step - loss: 0.5780 - accuracy: 0.7564
Epoch 6/100
20/20 [==============================] - 0s 1ms/step - loss: 0.5420 - accuracy: 0.8141
Epoch 7/100
20/20 [==============================] - 0s 1ms/step - loss: 0.5194 - accuracy: 0.7756
Epoch 8/100
20/20 [==============================] - 0s 1ms/step - loss: 0.4909 - accuracy: 0.7885
Epoch 9/100
20/20 [==============================] - 0s 1ms/step - loss: 0.4450 - accuracy: 0.8013
Epoch 10/100
20/20 [==============================] - 0s 1ms/step - loss: 0.4145 - accuracy: 0.8462
Epoch 11/

20/20 [==============================] - 0s 1ms/step - loss: 0.0071 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0067 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0064 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0061 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0063 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0057 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0057 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0053 - accuracy: 1.0000
Epoch 91/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0055 - accuracy: 1.0000
Epoch 92/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0049 - accuracy: 1.0000
Epoch 93/100


In [ ]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 1ms/step - loss: 0.8629 - accuracy: 0.7692


[0.8629363775253296, 0.7692307829856873]

Training Accuracy >>> Test Accuracy

### Model with Dropout Layer

In [ ]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7653 - accuracy: 0.4744
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6923 - accuracy: 0.5256
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7149 - accuracy: 0.4872
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7439 - accuracy: 0.4423
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6974 - accuracy: 0.5064
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7053 - accuracy: 0.5000
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7035 - accuracy: 0.4808
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6994 - accuracy: 0.4679
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6984 - accuracy: 0.5385
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6833 - accuracy: 0.5000
Epoch 11/

20/20 [==============================] - 0s 2ms/step - loss: 0.3426 - accuracy: 0.8590
Epoch 84/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3245 - accuracy: 0.9038
Epoch 85/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3396 - accuracy: 0.8526
Epoch 86/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3587 - accuracy: 0.8590
Epoch 87/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3705 - accuracy: 0.8205
Epoch 88/100
20/20 [==============================] - 0s 2ms/step - loss: 0.2835 - accuracy: 0.8654
Epoch 89/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3712 - accuracy: 0.8013
Epoch 90/100
20/20 [==============================] - 0s 1ms/step - loss: 0.3640 - accuracy: 0.8333
Epoch 91/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3402 - accuracy: 0.8974
Epoch 92/100
20/20 [==============================] - 0s 1ms/step - loss: 0.3445 - accuracy: 0.8654
Epoch 93/100


In [ ]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 1ms/step - loss: 0.3962 - accuracy: 0.8077


[0.39620405435562134, 0.807692289352417]

Training Accuracy is still good but Test Accuracy Improved

**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**